# 第一章 异常

## 1.1 异常概念

* **异常**：指的是程序在执行过程中，出现的非正常的情况，最终会导致JVM的非正常停止。

在Java等面向对象的编程语言中，异常本身是一个类，产生异常就是创建异常对象并抛出了一个异常对象。Java处理异常的方式是中断处理。

> 异常指的并不是语法错误,语法错了,编译不通过,不会产生字节码文件,根本不能运行。

## 1.2 异常体系

异常机制其实是帮助我们找到程序中的问题，异常的根类是`java.lang.Throwable`，其下有两个子类：`java.lang.Error`与`java.lang.Exception`，平常所说的异常指`java.lang.Exception`。

- ![异常体系](./images/异常体系.png)

**Throwable体系**

* **Error**：严重错误Error，无法通过处理的错误，只能事先避免。
* **Exception**:表示异常，异常产生后程序员可以通过代码的方式纠正，使程序继续运行，是必须要处理的。

**Throwable中的常用方法：**

* `public void printStackTrace()`：打印异常的详细信息。包含了异常的类型，异常的原因，异常出现的位置。
* `public String getMessage()`：获取发生异常的原因。提示给用户的时候，就提示错误原因。
* `public String toString()`：获取异常的类型和异常描述信息(不用)。

## 1.3 异常分类

**异常(Exception)的分类**

* **编译时期异常**：checked异常。在编译时期,就会检查,如果没有处理异常,则编译失败。(如日期格式化异常)
* **运行时期异常**：runtime异常。在运行时期,检查异常.在编译时期,运行异常不会编译器检测(不报错)。(如数学异常)。

- ![异常的分类](./images/异常的分类.png)

## 1.4 异常的处理

Java异常处理的五个关键字:**try、catch、finally、throw、throws**。

### 1.4.1 抛出异常throw

java中，提供了一个throw关键字，它用来抛出一个指定的异常对象。

1. 创建一个异常对象。封装一些提示信息(信息可以自己编写)。
2. throw**用在方法内**，用来抛出一个异常对象，将这个异常对象传递到调用者处，并结束当前方法的执行。

**格式：**

```java
throw new 异常类名(参数);
```

**示例：**

```java
throw new NullPointerException("要访问的arr数组不存在");

throw new ArrayIndexOutOfBoundsException("该索引在数组中不存在，已超出范围");
```

**示例：**

```java
public static void main(String[] args) {

    int[] arr = {2, 4, 52, 2};
    int index = 4;
    int element = getElement(arr, index);

    System.out.println(element);
    System.out.println("over");
}

    /*
    * 根据 索引找到数组中对应的元素
    */
public static int getElement(int[] arr, int index) {
    //判断 索引是否越界
    if (index < 0 || index > arr.length-1){
        /*
        判断条件如果满足，当执行完throw抛出异常对象后，方法已经无法继续运算。
        这时就会结束当前方法的执行，并将异常告知给调用者。这时就需要通过异常来解决。
        */
        throw new ArrayIndexOutOfBoundsException("索引越界了~~~");
    }
    int element = arr[index];
    return element;
}
```

### 1.4.2 Objects非空判断

* `public static <T> T requireNonNull(T obj)`：查看指定引用对象不是null。

### 1.4.3 声明异常throws

**声明异常**：将问题标识出来，报告给调用者。

关键字**throws**运用于方法声明之上,用于表示当前方法不处理异常,而是提醒该方法的调用者来处理异常(抛出异常)。

**格式:**

```java
修饰符 返回值类型 方法名(参数) throws 异常类名1,异常类名2...{ }
```

**示例：**

```java
    public static void main(String[] args) throws IOException {
        read("a.txt");
    }

    // 如果定义功能时有问题发生需要报告给调用者。可以通过在方法上使用throws关键字进行声明
    public static void read(String path) throws FileNotFoundException, IOException {

        if (!path.equals("b.txt")) {//如果不是a.txt这个文件
            // 如果不是 b.txt认为该文件不存在是一个错误 也就是异常 throw
            throw new FileNotFoundException("文件不存在");
        }
    }
```

### 1.4.5 捕获异常try...catch

1. 该方法不处理,而是声明抛出,由该方法的调用者来处理(throws)。
2. 在方法中使用try-catch的语句块来处理异常。

**try-catch**的方式就是捕获异常。
* **捕获异常**：Java中对异常有针对性的语句进行捕获，可以对出现的异常进行指定方式的处理。

**示例：**

```java
try{ 
    编写可能会出现异常的代码
}catch(异常类型 e){ 
    处理异常的代码
//记录日志/打印异常信息/继续抛出异常 
}
```

**try**：该代码块中编写可能产生异常的代码。
**catch**：用来进行某种异常的捕获，实现对捕获到的异常进行处理。

> Tips:
> try和catch都不能单独使用,必须连用。

**示例：**

```java
public static void main(String[] args) {

    try {// 当产生异常时，必须有处理方式。要么捕获，要么声明。
            read("b.txt");
    } catch (FileNotFoundException e) {// 括号中需要定义什么呢?
        // try中抛出的是什么异常，在括号中就定义什么异常类型
            System.out.println(e);
    }
    System.out.println("over");
}

/* *
    * 当前的这个方法中有异常有编译期异常
    */
public static void read(String path) throws FileNotFoundException {
    if (!path.equals("a.txt")) {//如果不是 a.txt这个文件
    //如果不是 a.txt 认为 该文件不存在 是一个错误 也就是异常 throw
    throw new FileNotFoundException("文件不存在"); }
}
```

Throwable类中定义了一些查看获取异常信息方法:

* `public String getMessage()`：获取异常的描述信息,原因(提示给用户的时候,就提示错误原因。
* `public String toString()`：获取异常的类型和异常描述信息(不用)。
* `public void printStackTrace()`：打印异常的跟踪栈信息并输出到控制台。

> 包括了异常的类型，异常的原因，还包括了异常出现的位置，在开发和调试阶段，都得使用printStackTrace。

### 1.4.6 finally 代码块

**finally**：有一些特定的代码无论异常是否发生，都需要执行。另外，因为异常会引发程序跳转，导致有些语句执行不到。而finally就是解决这个问题的，在finally代码块中存放的代码都是一定会被执行的。

**格式：**

```java
try...catch...finally:自身需要处理异常,最终还得关闭资源。
```

> Tips:
> finally不能单独使用。

**示例：**

```java
public static void main(String[] args) {
    try {
        read("a.txt");
    } catch (FileNotFoundException e) {
        //抓取到的是编译期异常 抛出去的是运行期
        throw new RuntimeException(e);
    } finally {
        System.out.println("不管程序怎样，这里都将会被执行。");
    }
    System.out.println("over");
}

/* *
    * 当前的这个方法中 有异常 有编译期异常
    */
public static void read(String path) throws FileNotFoundException {
    if (!path.equals("a.txt")) {//如果不是 a.txt这个文件
        //如果不是a.txt认为该文件不存在是一个错误也就是异常throw
        throw new FileNotFoundException("文件不存在");
    }
}
```

> Tips:
> 当只有在try或者catch中调用退出JVM的相关方法,此时finally才不会执行,否则finally永远会执行。

### 1.4.7 异常注意事项

1. 多个异常分别处理。
2. 多个异常一次捕获，多次处理。
3. 多个异常一次捕获一次处理。

一般我们是使用一次捕获多次处理方式。

**示例：**

```java
try{
    编写可能会出现异常的代码
}catch(异常类型A e){ 当try中出现A类型异常,就用该catch来捕获.
    处理异常的代码
    //记录日志/打印异常信息/继续抛出异常
}catch(异常类型B e){ 当try中出现B类型异常,就用该catch来捕获.
    处理异常的代码
    //记录日志/打印异常信息/继续抛出异常
}
```

> Tips：
> 这种异常处理方式，要求多个catch中的异常不能相同，并且若catch中的多个异常之间有子父类异常的关系，那么子类异常要求在上面的catch处理，父类异常在下面的catch处理。

* 运行时异常被抛出可以不处理。即不捕获也不声明抛出。
* 如果finally有return语句,永远返回finally中的结果,避免该情况。
* 如果父类抛出了多个异常,子类重写父类方法时,抛出和父类相同的异常或者是父类异常的子类或者不抛出异常。
* 父类方法没有抛出异常，子类重写父类该方法时也不可抛出异常。此时子类产生该异常，只能捕获处理，不能声明抛出。

## 1.5 自定义异常

**异常类如何定义:**

1. 自定义一个编译期异常：自定义类 并继承于`java.lang.Exception`。
2. 自定义一个运行时期的异常类：自定义类 并继承于`java.lang.RuntimeException`。

**示例：**

```java
// 业务逻辑异常
public RegisterException() {

}

public RegisterException(String message) {
    super(message);
}
```

```java
    // 模拟数据库中已存在账号
    private static String[] names = {"bill", "hill", "jill"};

    public static void main(String[] args) {

        //调用方法
        try {
            // 可能出现异常的代码
            checkUsername("bill");
            System.out.println("注册成功");//如果没有异常就是注册成功
        } catch (RegisterException e) {
            //处理异常
            e.printStackTrace();
        }
    }

    //判断当前注册账号是否存在
    //因为是编译期异常，又想调用者去处理 所以声明该异常
    public static boolean checkUsername(String uname) throws RegisterException {

        for (String name : names) {
            if (name.equals(uname)) {//如果名字在这里面 就抛出登陆异常
                throw new RegisterException("账号：" + name + "已经被注册了!");
            }
        }
        return true;
    }
```